AppCenterData/AppCenterDataTests/MSDataTests.m

```diff
@@ -2819,4 +2821,125 @@ - (void)testRemoteOperationDelegateMethodIsCalledWithNoExistenceOfImplementation
+  XCTAssertTrue([documentChanges[0].documentId isEqualToString:doc1Id]);
+  XCTAssertTrue([documentChanges[0].partition isEqualToString:doc1Ptn]);
+  XCTAssertEqual(documentChanges[0].timestamp, doc1Ts);
+  XCTAssertTrue([documentChanges[1].operation isEqualToString:doc2Op]);
+  XCTAssertTrue([documentChanges[1].documentId isEqualToString:doc2Id]);
+  XCTAssertTrue([documentChanges[1].partition isEqualToString:doc2Ptn]);
+  XCTAssertEqual(documentChanges[1].timestamp, doc2Ts);
+
+  OCMVerifyAll(dataMock);
+  [dataMock stopMocking];
+}
+
+- (void)testDocumentChangeDelegateMethodIsNotCalledWithoutDocumentChangeKey {
+  id dataMock = OCMPartialMock(self.sut);
+  OCMStub([dataMock sharedInstance]).andReturn(dataMock);
+  id dataDelegateMock = OCMProtocolMock(@protocol(MSDataDelegate));
+  OCMReject([dataDelegateMock data:self.sut didReceiveDocumentChangeNotification:OCMOCK_ANY]);
+  [MSData setDelegate:dataDelegateMock];
+
+  NSString *const documentChangesJson = @"[{\"op\":\"replace\",\"id\":\"BHGVGD5\",\"ptn\":\"user\",\"_ts\":1000}]";
+  NSDictionary<NSString *, NSString *> *notificationData = @{@"unknown-key" : documentChangesJson};
+
+  [dataMock appCenter:nil didReceiveServiceNotification:notificationData];
+
+  OCMVerifyAll(dataMock);
+  [dataMock stopMocking];
+}
+
+- (void)testDocumentChangeDelegateMethodIsNotCalledWithInvalidJson {
+  id dataMock = OCMPartialMock(self.sut);
+  OCMStub([dataMock sharedInstance]).andReturn(dataMock);
+  id dataDelegateMock = OCMProtocolMock(@protocol(MSDataDelegate));
+  OCMReject([dataDelegateMock data:self.sut didReceiveDocumentChangeNotification:OCMOCK_ANY]);
+  [MSData setDelegate:dataDelegateMock];
+
+  NSString *const documentChangesJson = @"this is not valid document change JSON";
+  NSDictionary<NSString *, NSString *> *notificationData = @{kMSServiceNotificationDocumentChangeKey : documentChangesJson};
+
+  [dataMock appCenter:nil didReceiveServiceNotification:notificationData];
+
+  OCMVerifyAll(dataMock);
+  [dataMock stopMocking];
+}
+
+- (void)testDocumentChangeDelegateMethodIsNotCalledWithNoChanges {
+  id dataMock = OCMPartialMock(self.sut);
+  OCMStub([dataMock sharedInstance]).andReturn(dataMock);
+  id dataDelegateMock = OCMProtocolMock(@protocol(MSDataDelegate));
+  OCMReject([dataDelegateMock data:self.sut didReceiveDocumentChangeNotification:OCMOCK_ANY]);
+  [MSData setDelegate:dataDelegateMock];
+
+  NSString *const documentChangesJson = @"[]";
+  NSDictionary<NSString *, NSString *> *notificationData = @{kMSServiceNotificationDocumentChangeKey : documentChangesJson};
+
+  [dataMock appCenter:nil didReceiveServiceNotification:notificationData];
+
+  OCMVerifyAll(dataMock);
+  [dataMock stopMocking];
+}
+
+- (void)testDocumentChangeDelegateMethodIsNotCalledForDelegateThatDoesNotRespondToSelector {
+  id dataMock = OCMPartialMock(self.sut);
+  OCMStub([dataMock sharedInstance]).andReturn(dataMock);
+  id dataDelegateMock = OCMPartialMock([MSMockDataDelegate new]);
```

Comment:
You can use protocol mock then you don't need `MSMockDataDelegate`.

Reply:
How do I use protocol mock and have it not implement the optional method? I added this because I couldn't find another way to have respondsToSelector return false.

Reply:
Ah.. OK. Then you can declare
```
@interface MSMockDataDelegate : NSObject <MSDataDelegate>
@end
```
in this file right below imports. I don't think `MSMockDataDelegate` is used by any other classes.

Reply:
No it isn't used anywhere else.

I have removed the files and added both the interface and implementation to the top of this file. I can't see a way to mock just the interface without declaring an implementation as it causes compile errors.